File: backend/app/schemas/apk_scan.py
"""Pydantic response schemas for APK security scanning endpoints."""

from __future__ import annotations

from pydantic import BaseModel, Field

# ---------------------------------------------------------------------------
# Phase 1: Manifest scan response schemas
# ---------------------------------------------------------------------------


class ManifestFindingResponse(BaseModel):
    """A single manifest security finding."""
    check_id: str
    title: str
    description: str
    severity: str
    evidence: str = ""
    cwe_ids: list[str] = Field(default_factory=list)
    confidence: str = "high"
    suppressed: bool = False
    suppression_reason: str = ""

    model_config = {"from_attributes": True}


class ManifestScanSummary(BaseModel):
    """Summary statistics for a manifest scan."""
    total_findings: int = 0
    critical: int = 0
    high: int = 0
    medium: int = 0
    low: int = 0
    info: int = 0


class ConfidenceSummary(BaseModel):
    """Confidence breakdown for scan findings."""
    high: int = 0
    medium: int = 0
    low: int = 0


class FirmwareContextResponse(BaseModel):
    """Firmware metadata context attached to scan responses."""
    device_model: str | None = None
    manufacturer: str | None = None
    android_version: str | None = None
    api_level: int | None = None
    security_patch: str | None = None
    architecture: str | None = None
    partition: str | None = None
    firmware_filename: str | None = None
    bootloader_state: str | None = None
    is_priv_app: bool = False
    is_system_app: bool = False
    is_vendor_app: bool = False

    model_config = {"from_attributes": True}


class ManifestScanResponse(BaseModel):
    """Full response for a manifest security scan."""
    package: str
    findings: list[ManifestFindingResponse] = Field(default_factory=list)
    summary: ManifestScanSummary = Field(default_factory=ManifestScanSummary)
    confidence_summary: ConfidenceSummary = Field(
        default_factory=ConfidenceSummary
    )
    is_priv_app: bool = False
    is_platform_signed: bool = False
    is_debug_signed: bool = False
    severity_reduced: bool = False
    reduced_check_ids: list[str] = Field(default_factory=list)
    suppressed_findings: list[ManifestFindingResponse] = Field(
        default_factory=list
    )
    suppressed_count: int = 0
    suppression_reasons: list[str] = Field(default_factory=list)
    from_cache: bool = False
    elapsed_ms: float | None = None
    error: str | None = None
    firmware_context: FirmwareContextResponse | None = None

    model_config = {"from_attributes": True}


# ---------------------------------------------------------------------------
# Phase 2a: Bytecode scan response schemas
# ---------------------------------------------------------------------------


class BytecodeFindingLocation(BaseModel):
    """A single location where a pattern was detected."""
    caller_class: str | None = None
    caller_method: str | None = None
    target: str | None = None
    string_value: str | None = None
    using_class: str | None = None
    using_method: str | None = None
    dangerous_class: str | None = None

    model_config = {"from_attributes": True}


class BytecodeFindingResponse(BaseModel):
    """A single bytecode security finding."""
    pattern_id: str
    title: str
    description: str
    severity: str
    confidence: str = "high"
    cwe_ids: list[str] = Field(default_factory=list)
    category: str
    locations: list[dict] = Field(default_factory=list)
    total_occurrences: int = 0

    model_config = {"from_attributes": True}


class BytecodeScanSummary(BaseModel):
    """Summary statistics for a bytecode scan."""
    total_findings: int = 0
    by_severity: dict[str, int] = Field(default_factory=dict)
    by_category: dict[str, int] = Field(default_factory=dict)
    by_confidence: dict[str, int] = Field(default_factory=dict)


class BytecodeScanResponse(BaseModel):
    """Full response for a bytecode scan."""
    package: str
    findings: list[BytecodeFindingResponse] = Field(default_factory=list)
    summary: BytecodeScanSummary = Field(default_factory=BytecodeScanSummary)
    elapsed_seconds: float = 0.0
    dex_count: int = 0
    from_cache: bool = False
    error: str | None = None
    firmware_context: FirmwareContextResponse | None = None

    model_config = {"from_attributes": True}


# ---------------------------------------------------------------------------
# Phase 2b: mobsfscan SAST scan response schemas
# ---------------------------------------------------------------------------


class SastFindingResponse(BaseModel):
    """A single SAST finding from mobsfscan."""
    rule_id: str
    title: str
    description: str
    severity: str
    file_path: str | None = None
    source_file: str | None = None  # Java/Kotlin source path for source viewer
    line_number: int | None = None
    cwe_ids: list[str] = Field(default_factory=list)
    owasp_mobile: str = ""
    masvs: str = ""

    model_config = {"from_attributes": True}


class SastScanTimingResponse(BaseModel):
    """Pipeline timing breakdown."""
    total_elapsed_ms: int = 0
    jadx_elapsed_ms: int = 0
    mobsfscan_elapsed_ms: int = 0


class SastScanSummary(BaseModel):
    """Summary statistics for a SAST scan."""
    total_findings: int = 0
    by_severity: dict[str, int] = Field(default_factory=dict)
    files_scanned: int = 0
    normalized_findings: int = 0
    persisted_count: int = 0
    suppressed_rule_count: int = 0
    suppressed_path_count: int = 0


class SastScanResponse(BaseModel):
    """Full response for a SAST (jadx+mobsfscan) scan."""
    success: bool = True
    findings: list[SastFindingResponse] = Field(default_factory=list)
    summary: SastScanSummary = Field(default_factory=SastScanSummary)
    timing: SastScanTimingResponse = Field(default_factory=SastScanTimingResponse)
    cached: bool = False
    error: str | None = None
    firmware_context: FirmwareContextResponse | None = None

    model_config = {"from_attributes": True}


# ---------------------------------------------------------------------------
# Decompiled source viewer
# ---------------------------------------------------------------------------


class SourceFileResponse(BaseModel):
    """Decompiled Java/Kotlin source code for a single file."""
    path: str
    source: str
    apk_path: str
    line_count: int


class SourceFileListResponse(BaseModel):
    """List of available decompiled source files for an APK."""
    apk_path: str
    files: list[str]
    total: int

File: frontend/src/types/apkScan.ts
/**
 * TypeScript types for APK security scan findings.
 *
 * These types mirror the Pydantic response schemas in
 * backend/app/routers/apk_scan.py for all three scan phases:
 *   - Phase 1: Manifest security checks (18 MobSF-equivalent)
 *   - Phase 2a: Bytecode insecure API pattern detection
 *   - Phase 2b: SAST via jadx + mobsfscan pipeline
 */

// ---------------------------------------------------------------------------
// Shared types
// ---------------------------------------------------------------------------

/** Firmware metadata context attached to scan responses. */
export interface FirmwareContextResponse {
  device_model: string | null
  manufacturer: string | null
  android_version: string | null
  api_level: number | null
  security_patch: string | null
  architecture: string | null
  partition: string | null
  firmware_filename: string | null
  bootloader_state: string | null
  is_priv_app: boolean
  is_system_app: boolean
  is_vendor_app: boolean
}

/** Severity levels for APK scan findings. */
export type ApkScanSeverity = 'critical' | 'high' | 'medium' | 'low' | 'info'

/** Confidence levels for manifest findings. */
export type ApkScanConfidence = 'high' | 'medium' | 'low'

// ---------------------------------------------------------------------------
// Phase 1: Manifest scan types
// ---------------------------------------------------------------------------

/** A single manifest security finding. */
export interface ManifestFindingResponse {
  check_id: string
  title: string
  description: string
  severity: string
  evidence: string
  cwe_ids: string[]
  confidence: string
}

/** Summary statistics for a manifest scan. */
export interface ManifestScanSummary {
  total_findings: number
  critical: number
  high: number
  medium: number
  low: number
  info: number
}

/** Confidence breakdown for scan findings. */
export interface ConfidenceSummary {
  high: number
  medium: number
  low: number
}

/** Full response for a manifest security scan. */
export interface ManifestScanResponse {
  package: string
  findings: ManifestFindingResponse[]
  summary: ManifestScanSummary
  confidence_summary: ConfidenceSummary
  is_priv_app: boolean
  is_platform_signed: boolean
  severity_reduced: boolean
  reduced_check_ids: string[]
  from_cache: boolean
  elapsed_ms: number | null
  error: string | null
  firmware_context: FirmwareContextResponse | null
}

// ---------------------------------------------------------------------------
// Phase 2a: Bytecode scan types
// ---------------------------------------------------------------------------

/** A single location where a bytecode pattern was detected. */
export interface BytecodeFindingLocation {
  caller_class: string | null
  caller_method: string | null
  target: string | null
  string_value: string | null
  using_class: string | null
  using_method: string | null
  dangerous_class: string | null
}

/** A single bytecode security finding. */
export interface BytecodeFindingResponse {
  pattern_id: string
  title: string
  description: string
  severity: string
  confidence: string
  cwe_ids: string[]
  category: string
  locations: Record<string, unknown>[]
  total_occurrences: number
}

/** Summary statistics for a bytecode scan. */
export interface BytecodeScanSummary {
  total_findings: number
  by_severity: Record<string, number>
  by_category: Record<string, number>
  by_confidence: Record<string, number>
}

/** Full response for a bytecode scan. */
export interface BytecodeScanResponse {
  package: string
  findings: BytecodeFindingResponse[]
  summary: BytecodeScanSummary
  elapsed_seconds: number
  dex_count: number
  from_cache: boolean
  error: string | null
  firmware_context: FirmwareContextResponse | null
}

// ---------------------------------------------------------------------------
// Phase 2b: SAST scan types
// ---------------------------------------------------------------------------

/** A single SAST finding from mobsfscan. */
export interface SastFindingResponse {
  rule_id: string
  title: string
  description: string
  severity: string
  file_path: string | null
  source_file: string | null
  line_number: number | null
  cwe_ids: string[]
  owasp_mobile: string
  masvs: string
}

/** Pipeline timing breakdown. */
export interface SastScanTimingResponse {
  total_elapsed_ms: number
  jadx_elapsed_ms: number
  mobsfscan_elapsed_ms: number
}

/** Summary statistics for a SAST scan. */
export interface SastScanSummary {
  total_findings: number
  by_severity: Record<string, number>
  files_scanned: number
  normalized_findings: number
  persisted_count: number
}

/** Full response for a SAST (jadx + mobsfscan) scan. */
export interface SastScanResponse {
  success: boolean
  findings: SastFindingResponse[]
  summary: SastScanSummary
  timing: SastScanTimingResponse
  cached: boolean
  error: string | null
  firmware_context: FirmwareContextResponse | null
}

// ---------------------------------------------------------------------------
// Combined scan result (all phases)
// ---------------------------------------------------------------------------

/** Combined result from running all three scan phases concurrently. */
export interface ApkFullScanResult {
  manifest: ManifestScanResponse | null
  bytecode: BytecodeScanResponse | null
  sast: SastScanResponse | null
}

/** Options for the combined scan. */
export interface ApkScanOptions {
  persistFindings?: boolean
  minSeverity?: string
  forceRescan?: boolean
  timeout?: number
}
